### PR TITLE
Spectrum

### PR DESCRIPTION
**Pull request by jolcese**
# What's new

- Spectrum Analyzer Plugin

# Verification 

- See under Plugins

# Checklist (For Reviewer)

- [ ] PR has description of feature/bug or link to Confluence/Jira task
- [ ] Description contains actions to verify feature/bugfix
- [ ] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: applications/spectrum_analyzer/spectrum_analyzer.h [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+#define NUM_CHANNELS 132
+typedef struct {
+    float		max_rssi;
+    uint8_t		max_rssi_dec;
+    uint8_t		max_rssi_channel;
+    ChannelInfo chan_table[NUM_CHANNELS];
+} SpectrumAnalyzerModel; 
+
+bool update_values_flag;
```
Comment (skotopes) [inline applications/spectrum_analyzer/spectrum_analyzer.h]: No global variables allowed. All needed variables must be incapsulated in dynamically allocated struct.

File: applications/spectrum_analyzer/spectrum_analyzer.h [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+uint8_t vscroll;
+uint8_t min_chan;
+uint8_t max_chan;
+
+typedef struct {
+    ValueMutex* model_mutex;
+    osMessageQueueId_t event_queue;
+} SpectrumAnalyzerContext;
+
+static const uint8_t radio_config[][2] = {
```
Comment (skotopes) [inline applications/spectrum_analyzer/spectrum_analyzer.h]: It's bad practice to place such things in headers.

File: applications/spectrum_analyzer/spectrum_analyzer.h [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+uint8_t band;
+uint8_t width;
+char max_hold;
+char height;
+uint8_t vscroll;
+uint8_t min_chan;
+uint8_t max_chan;
+
+typedef struct {
+    ValueMutex* model_mutex;
```
Comment (skotopes) [inline applications/spectrum_analyzer/spectrum_analyzer.h]: We planning to drop value mutex in future updates, please check music_player for examples

File: applications/spectrum_analyzer/spectrum_analyzer.c [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+                        break;
+                    }
+                }
+            } 
+        } else {
+            FURI_LOG_T("spectrum", "osMessageQueue: event timeout");
+            // event timeout
+        }
+        view_port_update(view_port);
+        release_mutex(&model_mutex, model);
```
Comment (skotopes) [inline applications/spectrum_analyzer/spectrum_analyzer.c]: release mutex, then update

File: applications/spectrum_analyzer/spectrum_analyzer.c [hunk 5]
```diff
@@ -0,0 +1,475 @@
+#include <furi.h>
+#include <furi_hal.h>
+#include <gui/gui.h>
+#include <input/input.h>
+#include <stdlib.h>
+#include "spectrum_analyzer.h"
+
+void spectrum_analyzer_draw_scale(Canvas* canvas, const SpectrumAnalyzerModel* model) {
+
+    #define FREQ_BOTTOM_Y 50
```
Comment (skotopes) [inline applications/spectrum_analyzer/spectrum_analyzer.c]: Defines in the code section is a bad pattern, please move it to the beginning of the file.

File: applications/spectrum_analyzer/spectrum_analyzer.c [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+    snprintf(temp_str, 18, "%u", tag_left);
+    canvas_draw_str_aligned(canvas, FREQ_START_X, 63, AlignCenter, AlignBottom, temp_str);
+    snprintf(temp_str, 18, "%u", tag_center);
+    canvas_draw_str_aligned(canvas, 128/2, 63, AlignCenter, AlignBottom, temp_str);
+    snprintf(temp_str, 18, "%u", tag_right);
+    canvas_draw_str_aligned(canvas, FREQ_START_X + FREQ_LENGTH_X - 1, 63, AlignCenter, AlignBottom, temp_str);
+}
+
+static void spectrum_analyzer_render_callback(Canvas* const canvas, void* ctx) {
+    const SpectrumAnalyzerModel* model = acquire_mutex((ValueMutex*)ctx, 25);
```
Comment (skotopes) [inline applications/spectrum_analyzer/spectrum_analyzer.c]: Please switch to osMutex, we planning to drop value mutex in future releases

File: applications/spectrum_analyzer/spectrum_analyzer.c [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+            if (model->chan_table[ch].ss > model->max_rssi_dec) {
+                model->max_rssi_dec = model->chan_table[ch].ss;
+                model->max_rssi = (model->chan_table[ch].ss / 2) - 138;
+                model->max_rssi_channel = ch;
+            }
+
+            furi_hal_subghz_idle();
+        }  
+
+        release_mutex(context->model_mutex, model);
```
Comment (skotopes) [inline applications/spectrum_analyzer/spectrum_analyzer.c]: Holding model for too long is going to lockup gui. 
Best option will be to generate data, then acquire mutex, then update, then release mutex.

Also there is a way to speedup work with transceiver: for time critical code acquire bus handle and then use cc1101 driver directly.

File: applications/spectrum_analyzer/spectrum_analyzer.c [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+        }
+
+        model->max_rssi_dec = 0;
+    
+        for (ch = 0; ch < NUM_CHANNELS-1; ch++) { 
+            // tune(ch);
+            furi_hal_subghz_set_frequency(model->chan_table[ch].frequency);
+
+            furi_hal_subghz_rx();
+            osDelay(3);
```
Comment (skotopes) [inline applications/spectrum_analyzer/spectrum_analyzer.c]: furi_hal_delay_ms uses osDelay inside

File: applications/spectrum_analyzer/spectrum_analyzer.c [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+        view_port_update(view_port);
+        release_mutex(&model_mutex, model);
+    }
+    
+    view_port_enabled_set(view_port, false);
+    gui_remove_view_port(gui, view_port);
+    furi_record_close("gui");
+    view_port_free(view_port);
+    
+    osThreadTerminate(analyzer_thread);
```
Comment (skotopes) [inline applications/spectrum_analyzer/spectrum_analyzer.c]: It's prohibited to use this function. It's extremely unsafe: terminating thread is going to leave system in inconsistent state that is impossible to predict and resolve.

Please use furi_thread. Also please check new version of music_player, i've removed this anti-pattern.

**Review by skotopes — CHANGES_REQUESTED** [review]

**Comment by skotopes** [comment]
Also code build is broken please fix all compiler errors and format sources

**Comment by jolcese** [comment]
Hi @skotopes . Thanks for all the feedback! Will work on that and resubmit.

File: applications/spectrum_analyzer/spectrum_analyzer.h (hunk 1 above)
Comment (jolcese) [reply applications/spectrum_analyzer/spectrum_analyzer.h]: Sure

File: applications/spectrum_analyzer/spectrum_analyzer.h (hunk 2 above)
Comment (jolcese) [reply applications/spectrum_analyzer/spectrum_analyzer.h]: Fixed

File: applications/spectrum_analyzer/spectrum_analyzer.c (hunk 4 above)
Comment (jolcese) [reply applications/spectrum_analyzer/spectrum_analyzer.c]: OK

File: applications/spectrum_analyzer/spectrum_analyzer.c (hunk 5 above)
Comment (jolcese) [reply applications/spectrum_analyzer/spectrum_analyzer.c]: OK

File: applications/spectrum_analyzer/spectrum_analyzer.c (hunk 8 above)
Comment (jolcese) [reply applications/spectrum_analyzer/spectrum_analyzer.c]: OK

File: applications/spectrum_analyzer/spectrum_analyzer.c (hunk 7 above)
Comment (jolcese) [reply applications/spectrum_analyzer/spectrum_analyzer.c]: Makes sense. 
Will research speeding up transceiver

File: applications/spectrum_analyzer/spectrum_analyzer.h (hunk 3 above)
Comment (jolcese) [reply applications/spectrum_analyzer/spectrum_analyzer.h]: Fixed

File: applications/spectrum_analyzer/spectrum_analyzer.c (hunk 6 above)
Comment (jolcese) [reply applications/spectrum_analyzer/spectrum_analyzer.c]: Switched to osMutex

File: applications/spectrum_analyzer/spectrum_analyzer.c (hunk 9 above)
Comment (jolcese) [reply applications/spectrum_analyzer/spectrum_analyzer.c]: using furi_thread

**Comment by jolcese** [comment]
@skotopes I refactored using the music player example. Let me know if you see other changes

**Comment by jolcese** [comment]
Hi @skotopes , not sure what I need to do for the submodule/protobuf check that's failing.

**Comment by skotopes** [comment]
I've tested app with external generator and frequency analyzer: some frequencies are undetected, other detected incorrectly.

What is missing right now and should be present to fix issue: overlapping when scanning. Also it make sense to profile antenna and use it when scanning.

**Comment by shmonoff** [comment]
Hey
Thanks for the plug in, I've just tested it
I've a suggestion - maybe it'll be better not to remove detected freqs from the screen as quick as it is, because between re-read/refresh cycles the signal could disappear. Or maybe list a log of all detected frequencies upon exit?

**Comment by skotopes** [comment]
Please reopen when you fix issues and let me know if you need any help.

**Pull request closed without merging**